Vault pass nonutf8 2.2 compat

abadger opened the pull request:
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
vault password handling.  Various files have implementation details

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.2
```

##### SUMMARY
    Retain vault password as bytes in 2.2
    
    Prior to 2.2.1, the vault password was read in as byes and then remained
    bytes all the way through the code.  A bug existed where bytes and text
    were mixed, leading to a traceback with non-ascii passwords.  In devel,
    this was fixed by changing the read in password to text type to match
    with our overall strategy of converting at the borders.  This was
    backported to stable-2.2 for the 2.2.1 release.
    
    On reflection, this should not have been backported as it causes
    passwords which were originally non-utf-8 to become utf-8.  People will
    then have their working 2.2.x vault files become in-accessible.
    
    this commit pipes bytes all the way through the system for vault
    password.  That way if a password is read in as a non-utf-8 character
    sequence, it will continue to work in 2.2.2+.  This change is only for
    the 2.2 branch, not for 2.3 and beyond.
    
    Why not everywhere?  The reason is that non-utf-8 passwords will cause
    problems when vault files are shared between systems or users.  If the
    password is read from the prompt and one user/machine has a latin1
    encoded locale while a second one has utf-8, the non-ascii password
    typed in won't match between machines.  Deal with this by making sure
    that when we encrypt the data, we always use valid utf-8.
    
Fixes #20398

alikins commented:
Haven't tested this yet, but the code changes LGTM.

There are submodule ref changes in that diff though.

abadger committed:
Retain vault password as bytes in 2.2

Prior to 2.2.1, the vault password was read in as byes and then remained
bytes all the way through the code.  A bug existed where bytes and text
were mixed, leading to a traceback with non-ascii passwords.  In devel,
this was fixed by changing the read in password to text type to match
with our overall strategy of converting at the borders.  This was
backported to stable-2.2 for the 2.2.1 release.

On reflection, this should not have been backported as it causes
passwords which were originally non-utf-8 to become utf-8.  People will
then have their working 2.2.x vault files become in-accessible.

this commit pipes bytes all the way through the system for vault
password.  That way if a password is read in as a non-utf-8 character
sequence, it will continue to work in 2.2.2+.  This change is only for
the 2.2 branch, not for 2.3 and beyond.

Why not everywhere?  The reason is that non-utf-8 passwords will cause
problems when vault files are shared between systems or users.  If the
password is read from the prompt and one user/machine has a latin1
encoded locale while a second one has utf-8, the non-ascii password
typed in won't match between machines.  Deal with this by making sure
that when we encrypt the data, we always use valid utf-8.

Fixes #20398

abadger commented:
Good spot.  Fixed.

mattclay commented:
CI failure due to:

```
2017-02-15 22:33:35 ERROR! Unexpected Exception: __init__() got an unexpected keyword argument 'password'
```
